File: plugins/cpu.10s.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::thread;
use std::time::Duration;

fn main() {
    let cpu_percent = get_cpu_usage();

    // Determine color based on usage
    let color = if cpu_percent > 80.0 {
        "red"
    } else if cpu_percent > 50.0 {
        "yellow"
    } else {
        "green"
    };

    println!("\u{1F4BB} {:.1}% | color={}", cpu_percent, color);
    println!("---");
    println!("CPU Usage: {:.1}%", cpu_percent);

    // Get number of CPUs
    if let Ok(file) = File::open("/proc/cpuinfo") {
        let reader = BufReader::new(file);
        let cores = reader.lines()
            .filter_map(|l| l.ok())
            .filter(|l| l.starts_with("processor"))
            .count();
        println!("Cores: {}", cores);
    }

    println!("---");
    println!("Refresh | refresh=true");
}

fn get_cpu_usage() -> f64 {
    let stat1 = read_cpu_stat();
    thread::sleep(Duration::from_millis(100));
    let stat2 = read_cpu_stat();

    if stat1.is_empty() || stat2.is_empty() {
        return 0.0;
    }

    let idle1 = stat1.get(3).copied().unwrap_or(0);
    let idle2 = stat2.get(3).copied().unwrap_or(0);

    let total1: u64 = stat1.iter().sum();
    let total2: u64 = stat2.iter().sum();

    let total_delta = total2.saturating_sub(total1);
    let idle_delta = idle2.saturating_sub(idle1);

    if total_delta == 0 {
        return 0.0;
    }

    100.0 * (total_delta - idle_delta) as f64 / total_delta as f64
}

fn read_cpu_stat() -> Vec<u64> {
    if let Ok(file) = File::open("/proc/stat") {
        let reader = BufReader::new(file);
        for line in reader.lines().filter_map(|l| l.ok()) {
            if line.starts_with("cpu ") {
                return line.split_whitespace()
                    .skip(1)
                    .filter_map(|s| s.parse().ok())
                    .collect();
            }
        }
    }
    Vec::new()
}

File: plugins/cpu/cpu.10s.go
// +build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
)

func main() {
	cpuPercent := getCPUUsage()

	// Determine color based on usage
	var color string
	switch {
	case cpuPercent > 80:
		color = "red"
	case cpuPercent > 50:
		color = "yellow"
	default:
		color = "green"
	}

	fmt.Printf("\U0001F4BB %.1f%% | color=%s\n", cpuPercent, color)
	fmt.Println("---")
	fmt.Printf("CPU Usage: %.1f%%\n", cpuPercent)
	fmt.Printf("Cores: %d\n", runtime.NumCPU())
	fmt.Printf("Architecture: %s\n", runtime.GOARCH)
	fmt.Println("---")
	fmt.Println("Refresh | refresh=true")
}

func getCPUUsage() float64 {
	if runtime.GOOS == "linux" {
		return getLinuxCPU()
	}
	// Fallback for other OS
	return 0.0
}

func getLinuxCPU() float64 {
	stat1 := readCPUStat()
	time.Sleep(100 * time.Millisecond)
	stat2 := readCPUStat()

	idle1 := stat1[3]
	idle2 := stat2[3]

	total1 := sum(stat1)
	total2 := sum(stat2)

	totalDelta := total2 - total1
	idleDelta := idle2 - idle1

	if totalDelta == 0 {
		return 0.0
	}

	return 100.0 * float64(totalDelta-idleDelta) / float64(totalDelta)
}

func readCPUStat() []int64 {
	file, err := os.Open("/proc/stat")
	if err != nil {
		return make([]int64, 10)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "cpu ") {
			fields := strings.Fields(line)[1:]
			stats := make([]int64, len(fields))
			for i, f := range fields {
				val, _ := strconv.ParseInt(f, 10, 64)
				stats[i] = val
			}
			return stats
		}
	}
	return make([]int64, 10)
}

func sum(arr []int64) int64 {
	var total int64
	for _, v := range arr {
		total += v
	}
	return total
}
